packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase3Algorithm_kokkos_def.hpp

```diff
@@ -65,88 +65,123 @@ namespace MueLu {
+
+                               if (neigh != i && graph.isLocalNeighborVertex(neigh)
+                                   && aggStatView(neigh) == READY) {
+                                 isNewAggregate = true;
+
+                                 aggStatView (neigh)    = AGGREGATED;
+                                 vertex2AggId(neigh, 0) = numLocalAggregatesView();
+                                 procWinner  (neigh, 0) = myRank;
+
+                                 numNonAggregatedNodesView() = numNonAggregatedNodesView() - 1;
+                               }
+                             }
+
+                             if (isNewAggregate) {
+                               // Create new aggregate (not singleton)
+                               // aggregates.SetIsRoot(i);
+                               vertex2AggId(i, 0) = numLocalAggregatesView();
+                               numLocalAggregatesView() = numLocalAggregatesView() + 1;
+
+                             } else {
+                               // We do not want a singleton, but there are no non-aggregated
+                               // neighbors. Lets see if we can connect to any other aggregates
+                               // NOTE: This is very similar to phase 2b, but simpler: we stop with
+                               // the first found aggregate
+                               int j = 0;
+                               for (; j < as<int>(neighOfINode.length); ++j) {
+                                 LO neigh = neighOfINode(j);
+
+                                 // We don't check (neigh != rootCandidate), as it is covered by checking (aggStatView(neigh) == AGGREGATED)
+                                 if (graph.isLocalNeighborVertex(neigh)
+                                     && aggStatView(neigh) == AGGREGATED)
+                                   break;
+                               }
+
+                               if (j < as<int>(neighOfINode.length)) {
+                                 // Assign to an adjacent aggregate
+                                 vertex2AggId(i, 0) = vertex2AggId(neighOfINode(j), 0);
+                               } else if (error_on_isolated) {
+                                 // Error on this isolated node, as the user has requested
+                                 std::ostringstream oss;
+                                 oss<<"MueLu::AggregationPhase3Algorithm::BuildAggregates: MueLu has detected a non-Dirichlet node that has no on-rank neighbors and is terminating (by user request). "<<std::endl;
+                                 oss<<"If this error is being generated at level 0, this is due to an initial partitioning problem in your matrix."<<std::endl;
+                                 oss<<"If this error is being generated at any other level, try turning on repartitioning, which may fix this problem."<<std::endl;
+                                 throw Exceptions::RuntimeError(oss.str());
```

Comment:
Throwing in device code is giving me a compile error with CUDA. Would it be possible to enable the error checking and output in host space only? I'm not sure of the most elegant way to do that.